cmd/oras/root/push.go

```diff
@@ -116,7 +122,7 @@ Example - Push file "hi.txt" into an OCI image layout folder 'layout-dir' with t
 				return err
 			}
 
-			if opts.manifestConfigRef != "" && opts.artifactType == "" {
+			if (opts.manifestConfigRef != "" || opts.Platform.Platform != nil) && opts.artifactType == "" {
```

Comment:
This is need if we are going to allow v1.1 `application/vnd.unknown.artifact.v1` artifact to have platform information in config blob.

Reply:
@shizhMSFT If `--platform` is used without `--artifact-type` and `image-spec` is not provided, should we pack v1.1 or v1.0 image?

Reply:
Currently if `--config` is not provided and `image-spec` is not provided, `oras` will pack v1.1 artifact
```console
➜  ~ oras push localhost:5000/test:type
✓ Uploaded  application/vnd.oci.empty.v1+json                                                                   2/2  B 100.00%   34ms
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      535/535  B 100.00%     0s
  └─ sha256:5f56fc622e5f67d225a27be2c3637a606880c23aa98a2561b3797c4b54ac9fd5
Pushed [registry] localhost:5000/test:type
ArtifactType: application/vnd.unknown.artifact.v1
Digest: sha256:5f56fc622e5f67d225a27be2c3637a606880c23aa98a2561b3797c4b54ac9fd5
➜  ~ oras manifest fetch --pretty localhost:5000/test@sha256:5f56fc622e5f67d225a27be2c3637a606880c23aa98a2561b3797c4b54ac9fd5
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.unknown.artifact.v1",
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2,
    "data": "e30="
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.empty.v1+json",
      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "size": 2,
      "data": "e30="
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-10-10T06:29:05Z"
  }
}
```

Reply:
If `--config` is used with `--artifact-type`, `oras` will also pack v1.1 artifact
```console
➜  ~ oras push localhost:5000/test:v1.1 --artifact-type test/123 --config 123
✓ Exists    application/vnd.oci.empty.v1+json                                                                   2/2  B 100.00%     0s
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.unknown.config.v1+json                                                              3/3  B 100.00%   39ms
  └─ sha256:ca3d163bab055381827226140568f3bef7eaac187cebd76878e0b63e9e442356
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      499/499  B 100.00%     0s
  └─ sha256:d80e321ea40ab36b661693ddf507f425801819481949d6bdbb6ec52f503139ae
Pushed [registry] localhost:5000/test:v1.1
ArtifactType: test/123
Digest: sha256:d80e321ea40ab36b661693ddf507f425801819481949d6bdbb6ec52f503139ae
➜  ~ oras manifest fetch --pretty localhost:5000/test@sha256:d80e321ea40ab36b661693ddf507f425801819481949d6bdbb6ec52f503139ae
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "test/123",
  "config": {
    "mediaType": "application/vnd.unknown.config.v1+json",
    "digest": "sha256:ca3d163bab055381827226140568f3bef7eaac187cebd76878e0b63e9e442356",
    "size": 3
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.empty.v1+json",
      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "size": 2,
      "data": "e30="
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-10-10T06:34:43Z"
  }
}
```

If config media type is not provided, it will be set to `application/vnd.unknown.config.v1+json`. We can follow this practice for `--platform` and bake v1.1 artifact for `oras push --platform xxx --artifact xxx` usage.

Reply:
> @shizhMSFT If `--platform` is used without `--artifact-type` and `image-spec` is not provided, should we pack v1.1 or v1.0 artifact? (I would prefer v1.1)

Let's follow the default value of `image-spec` when this flag is not provided.

Reply:
We make `oras push --config xxx` craft v1.0 artifact to avoid breaking change. `oras push --platform xxx` should create v1.1 artifact. The suggested change should be applied.

Reply:
v1.1 image spec is default except for that case you mention `--config` and no `--artifact-type`

https://github.com/oras-project/oras/blob/main/cmd/oras/root/push.go#L122-L128

Reply:
Yes, and we should make v1.1 default when `--artifact-platform xxx` is used regardless of `--artifact-type`.

Having `|| opts.Platform.Platform != nil` here will make v1.0 default when `--artifact-platform` used without `--artifact-type`.

Reply:
Done